File: scripts/generate-pdfs.mjs
import http from "node:http";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { chromium } from "playwright";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(__dirname, "..");
const PDF_DIR = path.join(ROOT, "pdf");
const PORT = 4173;

const GENES = [
  "ACTN3", "ADRA2A", "ANK3", "ANKK1", "APOE", "BDNF", "CACNA1C", "CDH13",
  "CHRNA5", "CLOCK", "COMT", "CYP1A2", "DBH", "DRD2", "FKBP5", "FTO",
  "LCT", "MAOA", "MC1R", "MTHFR", "OXTR", "SLC6A4", "TAS2R38", "TPH2",
];

const MIME = {
  ".html": "text/html; charset=utf-8",
  ".css": "text/css; charset=utf-8",
  ".js": "text/javascript; charset=utf-8",
  ".md": "text/plain; charset=utf-8",
  ".json": "application/json; charset=utf-8",
};

function startStaticServer() {
  return new Promise((resolve) => {
    const server = http.createServer((req, res) => {
      try {
        const url = new URL(req.url || "/", `http://127.0.0.1:${PORT}`);
        let rel = decodeURIComponent(url.pathname);
        if (rel === "/") {
          rel = "/index.html";
        }
        const safe = path.normalize(rel).replace(/^(\.\.(\/|\\|$))+/, "");
        const filePath = path.join(ROOT, safe.replace(/^\//, ""));
        if (!filePath.startsWith(ROOT) || !fs.existsSync(filePath) || fs.statSync(filePath).isDirectory()) {
          res.writeHead(404);
          res.end("Not found");
          return;
        }
        const ext = path.extname(filePath).toLowerCase();
        res.writeHead(200, { "Content-Type": MIME[ext] || "application/octet-stream" });
        fs.createReadStream(filePath).pipe(res);
      } catch {
        res.writeHead(500);
        res.end("Server error");
      }
    });
    server.listen(PORT, "127.0.0.1", () => resolve(server));
  });
}

async function waitForPrintReady(page) {
  await page.waitForFunction(
    () => {
      const pages = document.querySelectorAll(".print-document .print-page");
      return pages.length === 4 && pages[0].querySelector(".section-card");
    },
    { timeout: 30000 }
  );
  await page.evaluate(async () => {
    if (typeof window.fitPrintPagesToSheet === "function") {
      window.fitPrintPagesToSheet();
    }
    await new Promise((resolve) => requestAnimationFrame(() => requestAnimationFrame(resolve)));
    if (typeof window.fitPrintPagesToSheet === "function") {
      window.fitPrintPagesToSheet();
    }
  });
  await page.waitForTimeout(200);
}

async function generatePdf(browser, gene) {
  const page = await browser.newPage();
  const url = `http://127.0.0.1:${PORT}/html/print.html?gene=${encodeURIComponent(gene)}`;

  try {
    await page.goto(url, { waitUntil: "networkidle" });
    await waitForPrintReady(page);

    const outPath = path.join(PDF_DIR, `${gene}.pdf`);
    await page.pdf({
      path: outPath,
      format: "A4",
      printBackground: true,
      preferCSSPageSize: true,
      margin: { top: "10mm", right: "10mm", bottom: "10mm", left: "10mm" },
    });
    console.log(`OK  ${gene}.pdf`);
    return true;
  } catch (error) {
    console.error(`ERR ${gene}: ${error.message}`);
    return false;
  } finally {
    await page.close();
  }
}

async function main() {
  fs.mkdirSync(PDF_DIR, { recursive: true });

  const server = await startStaticServer();
  const browser = await chromium.launch({ headless: true });

  let ok = 0;
  let fail = 0;

  try {
    console.log(`Generowanie ${GENES.length} PDF do: ${PDF_DIR}\n`);
    for (const gene of GENES) {
      const success = await generatePdf(browser, gene);
      if (success) {
        ok += 1;
      } else {
        fail += 1;
      }
    }
  } finally {
    await browser.close();
    server.close();
  }

  console.log(`\nGotowe: ${ok} sukces, ${fail} błędów.`);
  process.exit(fail > 0 ? 1 : 0);
}

main();

File: scripts/generate-pdfs.py
#!/usr/bin/env python3
"""Generuje karty genów jako PDF (4 strony, układ z html/print.html)."""

from __future__ import annotations

import http.server
import socketserver
import threading
import time
from pathlib import Path

from playwright.sync_api import sync_playwright

ROOT = Path(__file__).resolve().parent.parent
PDF_DIR = ROOT / "pdf"
PORT = 4173

GENES = [
    "ABCC11", "ACTN3", "ADRA2A", "ALDH2", "ANK3", "ANKK1", "APOE", "AR",
    "BDNF", "CACNA1C", "CDH13", "CHRNA5", "CLOCK", "COMT", "CYP1A2", "DBH",
    "DRD2", "FKBP5", "FTO", "GC", "HERC2", "LCT", "MAOA", "MC1R", "MTHFR",
    "OCA2", "OR2M", "OR6A2", "OXTR", "SLC24A4", "SLC45A2", "SLC6A4", "TAS2R38", "TPH2",
    "ZEB2",
]

WAIT_READY_JS = """
() => {
  const pages = document.querySelectorAll('.print-document .print-page');
  return pages.length === 4 && pages[0].querySelector('.section-card');
}
"""

FIT_JS = """
async () => {
  if (typeof window.fitPrintPagesToSheet === 'function') {
    window.fitPrintPagesToSheet();
  }
  await new Promise((r) => requestAnimationFrame(() => requestAnimationFrame(r)));
  if (typeof window.fitPrintPagesToSheet === 'function') {
    window.fitPrintPagesToSheet();
  }
}
"""


class QuietHandler(http.server.SimpleHTTPRequestHandler):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, directory=str(ROOT), **kwargs)

    def log_message(self, format, *args):
        pass


def start_server() -> socketserver.TCPServer:
    server = socketserver.TCPServer(("127.0.0.1", PORT), QuietHandler)
    thread = threading.Thread(target=server.serve_forever, daemon=True)
    thread.start()
    return server


def wait_for_ready(page) -> None:
    page.wait_for_function(WAIT_READY_JS, timeout=30_000)
    page.evaluate(FIT_JS)
    page.wait_for_timeout(200)


def generate_pdf(page, gene: str) -> bool:
    url = f"http://127.0.0.1:{PORT}/html/print.html?gene={gene}"
    out_path = PDF_DIR / f"{gene}.pdf"
    try:
        page.goto(url, wait_until="networkidle", timeout=60_000)
        wait_for_ready(page)
        page.pdf(
            path=str(out_path),
            format="A4",
            print_background=True,
            prefer_css_page_size=True,
            margin={"top": "10mm", "right": "10mm", "bottom": "10mm", "left": "10mm"},
        )
        print(f"OK  {gene}.pdf")
        return True
    except Exception as exc:
        print(f"ERR {gene}: {exc}")
        return False


def main() -> int:
    PDF_DIR.mkdir(parents=True, exist_ok=True)
    server = start_server()
    time.sleep(0.3)

    ok = 0
    fail = 0

    print(f"Generowanie {len(GENES)} PDF do: {PDF_DIR}\n")

    with sync_playwright() as playwright:
        browser = playwright.chromium.launch(headless=True)
        context = browser.new_context()
        page = context.new_page()

        try:
            for gene in GENES:
                if generate_pdf(page, gene):
                    ok += 1
                else:
                    fail += 1
        finally:
            context.close()
            browser.close()
            server.shutdown()

    print(f"\nGotowe: {ok} sukces, {fail} bledow.")
    return 1 if fail else 0


if __name__ == "__main__":
    raise SystemExit(main())
